File: src/shared/types/workspace.ts
import type { GameModSummary, GameScanWarning, RowData, TableKey } from './index';

/** Metadata for one imported Mod in the workspace */
export interface ModEntry {
  modRoot: string;
  displayName: string;
  version: string;
  status: 'loading' | 'ready' | 'error';
  error?: string;
}

/** Per-Mod table state — fully isolated from other Mods */
export interface ModTableState {
  tables: Record<TableKey, RowData[]>;
  originalTables: Record<TableKey, RowData[]>;
  dirty: Record<TableKey, Record<string, Record<string, string>>>;
  currentTab: TableKey;
  currentFaction: string;
  searchText: string;
  selectedRowKey: string;
  editing: { tab: TableKey; rowKey: string; col: string; value: string } | null;
  nextRowKey: number; // Per-mod row key counter to prevent collisions
}

/** Which main-content view is shown */
export type WorkspaceView = 'overview' | 'table' | 'settings' | 'config';

/** Which config sub-view is shown when currentView === 'config' */
export type ConfigView = 'mod-overview' | 'file-history' | 'mod-info' | 'factions' | 'variants' | 'mission';

/** Persisted mod entry (stored to disk) */
export interface PersistedMod {
  modRoot: string;
  displayName: string;
  version: string;
}

/** Persisted workspace state (stored to disk via Rust backend) */
export interface PersistedWorkspace {
  mods: PersistedMod[];
  activeModRoot: string | null;
  currentView: string | null;
  expandedMods: string[];
  starsectorRoot?: string | null;
  gameMods?: GameModSummary[];
  gameWarnings?: GameScanWarning[];
}

File: src-tauri/src/models/workspace.rs
use serde::{Deserialize, Serialize};

use super::{GameModSummary, GameScanWarning};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum WorkspaceView {
    Overview,
    Table,
    Settings,
    Config,
    About,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct PersistedWorkspace {
    #[serde(default)]
    pub mods: Vec<PersistedMod>,
    pub active_mod_root: Option<String>,
    pub current_view: Option<WorkspaceView>,
    #[serde(default)]
    pub expanded_mods: Vec<String>,
    pub starsector_root: Option<String>,
    #[serde(default)]
    pub game_mods: Vec<GameModSummary>,
    #[serde(default)]
    pub game_warnings: Vec<GameScanWarning>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PersistedMod {
    pub mod_root: String,
    pub display_name: String,
    #[serde(default)]
    pub version: String,
}
